p2p/host/relay/autorelay.go

```diff
@@ -253,6 +318,54 @@ func (ar *AutoRelay) doUpdateAddrs() {
 	ar.addrs = raddrs
 }
 
+// This function cleans up a relay's address set to remove private addresses and curtail
+// addrsplosion. For the latter, we use the following heuristic:
+// - if the address set includes a (tcp) address with the default port 4001,
+//   we remove all tcp addrs with a different port
+func cleanupAddressSet(pi pstore.PeerInfo) pstore.PeerInfo {
```

Comment:
What about stuff like `/ip4/.../tcp/4002/ws`? What about different IP addresses? Let's allow one port per public IP + transport (ws, quic, raw tcp, etc.).

We can also look at, e.g., the localhost advertisement (if it exists) to try to pick the best (non-standard) port.

Reply:
Ok, I'll concoct a better heuristic.

Reply:
I reworked the heuristic and moved it to its own file.
It works as follows:
- it declares addrsplosion if there are multiple ports for some base-address/protocol combination
- for each base-address/protocol combination with addrsplosion it selects the addrs where the port is 4001 or 4002 or it matches a privately bound port
- if we can't select any default or private port, bite the bullet and accept them all.